clang/docs/SYCLFERFCLoweringKernel.md

```diff
@@ -0,0 +1,93 @@
+Discourse topic details
+
+- Category: [Clang Frontend](https://discourse.llvm.org/c/clang/6)
+- Title: "RFC: SYCL kernel lowering"
+- Tags: sycl
+
+# SYCL Kernel Lowering
+
+A SYCL construct such as `parallel_for` or `single_task` takes a named function object or a lambda as one of its arguments.   The contents of this function object are executed on the device.   However, as the SYCL runtime can rely on other offload APIs like OpenCL or CUDA to execute the function object, it needs to respect the calling convention of these API.  To enable this, the function object is converted into the format of an OpenCL kernel.
+
+Consider the following code snippet:
+
+```
+#include <sycl.hpp>
+
+int main() {
+  sycl::queue q;
+  sycl::buffer<char> b{sycl::range{1024}};
+
+  q.submit([&](sycl::handler &cgh) {
+    sycl::accessor acc{b, cgh};
+    int i;
+    struct S {
+      char c;
+      int i;
+    } test_s;
+    test_s.c = 14;
+
+    cgh.single_task([=] {
+      if (i == 13 && test_s.c == 14) {
+        acc[0] = 'a';
+      }
+    });
+  });
+}
+```
+
+In this example, the lambda passed to the `single_task` construct needs to be executed on the device.  The corresponding function object looks like:
+
+```
+struct FuncObj {
+  int i;
+  struct S test_s;
+  sycl::accessor acc1;
+
+  void operator () {  // Function call operator
+    if (i == 13 && test_s.c == 14) {
+      acc[0] = 'a';
+    }
+  }
+};
+```
+
+The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.  It does so by walking the function object data member and generating a parameter for each of them.  Some special types like accessor are treated a bit differently (see below).  Inside the OpenCL kernel, the function object is rebuilt and then called.
+
+The device compiler transforms this into (pseudo-code):
+
+```
+    spir_kernel void Caller(
```

Comment:
Can we avoid references to SPIR or SPIR-V so as to avoid any appearance of a bias?